kanidmd/src/lib/idm/oauth2.rs

```diff
@@ -0,0 +1,1021 @@
+            client_name: auth_req.client_id.clone(),
+            scopes: Vec::new(),
+            consent_token,
+        })
+    }
+
+    pub fn check_oauth2_authorise_permit(
+        &self,
+        audit: &mut AuditScope,
+        ident: &Identity,
+        uat: &UserAuthToken,
+        consent_token: &str,
+        ct: Duration,
+    ) -> Result<AuthorisePermitSuccess, OperationError> {
+        // Decode the consent req with our system fernet key. Use a ttl of 5 minutes.
+        let consent_req: ConsentToken = self
+            .inner
+            .fernet
+            .decrypt_at_time(consent_token, Some(300), ct.as_secs())
+            .map_err(|_| {
+                ladmin_error!(audit, "Failed to decrypt consent request");
+                OperationError::CryptographyError
+            })
+            .and_then(|data| {
+                serde_json::from_slice(&data).map_err(|e| {
+                    ladmin_error!(audit, "Failed to deserialise consent request - {:?}", e);
+                    OperationError::SerdeJsonError
+                })
+            })?;
+
+        // Validate that the ident_id matches our current ident.
+        if consent_req.ident_id != ident.get_event_origin_id() {
+            lsecurity!(
+                audit,
+                "consent request ident id does not match the identity of our UAT."
+            );
+            return Err(OperationError::InvalidSessionState);
+        }
+
+        // Validate that the session id matches our uat.
+        if consent_req.session_id != uat.session_id {
+            lsecurity!(
+                audit,
+                "consent request sessien id does not match the session id of our UAT."
+            );
+            return Err(OperationError::InvalidSessionState);
+        }
+
+        // Get the resource server config based on this client_id.
+        let o2rs_fernet = match self.inner.rs_set.get(&consent_req.client_id) {
+            Some(Oauth2RS::Basic(rsbasic)) => &rsbasic.token_fernet,
+            None => {
+                ladmin_error!(audit, "Invalid consent request oauth2 client_id");
+                return Err(OperationError::InvalidRequestState);
+            }
+        };
+
+        // Extract the state, code challenge, redirect_uri
+
+        let xchg_code = TokenExchangeCode {
+            uat: uat.clone(),
+            code_challenge: consent_req.code_challenge,
+            redirect_uri: consent_req.redirect_uri.clone(),
+        };
+
+        // Encrypt the exchange token with the fernet key of the client resource server
+        let code_data = serde_json::to_vec(&xchg_code).map_err(|e| {
+            ladmin_error!(audit, "Unable to encode xchg_code data {:?}", e);
+            OperationError::SerdeJsonError
+        })?;
+
+        let code = o2rs_fernet.encrypt_at_time(&code_data, ct.as_secs());
+
+        Ok(AuthorisePermitSuccess {
+            redirect_uri: consent_req.redirect_uri,
+            state: consent_req.state,
+            code,
```

Comment:
So we are encrypting the `code` so that a (malicious) client cannot just generate his own `code` and `code_verifier` to get access token (when calling `idms_prox_read.check_oauth2_token_exchange`)?

Reply:
Yep, exactly. The only way you can have a valid "code" is if the user authorised and consented to the authorisation proceeding, and is within the session time window. A resource server can't "fake" a code for a user :)